crates/containerd-shim-wasm/src/sandbox/stdio.rs

```diff
@@ -0,0 +1,73 @@
+#[cfg_attr(unix, path = "sys/unix/stdio.rs")]
+#[cfg_attr(windows, path = "sys/windows/stdio.rs")]
+mod stdio_impl;
+
+use std::fs::{File, OpenOptions};
+use std::io::ErrorKind::NotFound;
+use std::io::{Error, Result};
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+
+use stdio_impl::*;
+
+#[derive(Default, Clone)]
+pub struct Stdio {
+    pub stdin: Stdin,
+    pub stdout: Stdout,
+    pub stderr: Stderr,
+}
+
+impl Stdio {
+    pub fn redirect(&self) -> Result<()> {
+        self.stdin.redirect()?;
+        self.stdout.redirect()?;
+        self.stderr.redirect()?;
+        Ok(())
+    }
+}
+
+macro_rules! stdio_impl {
+    ( $stdio_type:ident, $fd:expr ) => {
+        #[derive(Default, Clone)]
+        pub struct $stdio_type(Arc<Mutex<Option<File>>>);
```

Comment:
question: why do we want a Arc on the stdio?

Reply:
I believe this is because it's not clone/copy so won't work across both of the executors (https://github.com/containerd/runwasi/pull/240#issuecomment-1680836865)

Reply:
The `Stdio` struct has to be passed to the wasm executor and the linux / default executor. Both will have ownership of the files, but only one of them will end up using them.
In any case, `Arc` allows it to be `clone`d cheaply.

Reply:
Ahh I see! Is `Rc<RefCell<T>>` good enough since we aren't dealing with multithreading?

Reply:
That might work if it doesn't break some `Sync` requirements. I'll try it tomorrow.